design/open/spark-outputcommitter/committer.md

```diff
@@ -0,0 +1,148 @@
+  regardless of other writers used.
+* **Clearly correct by construction**: Rather than rely on single atomic
+  operations and carefully tailoring operations to Spark retry mechanisms,
+  we use lakeFS capabilities and guarantees.  Analyzing correctness becomes
+  simpler.
+* **Fast**: No data copies, just only (required) metadata operations.  Cost
+  of commit is linear in the number of objects it touches.  Total time to
+  write is close to 3* faster than the existing FileOutputCommitter in v1
+  mode, close to 2* faster than the existing FileOutputCommitter in v2 mode
+  (which is unsafe in various cases), and about as fast as the magic
+  OutputCommitter _if_ lakeFSFS supported it.
+* **Good semantics**: Spark commits will be lakeFS commits.  The history of
+  a Spark job appears right in lakeFS history.  Metadata even includes some
+  data lineage -- and in future we can easily add more, for instance as
+  merge commit user metadata.
+
+### Implementation details
+
+#### Hadoop >=3.1
+
+Hadoop 3.1 offered a fairly complete overhaul of committer architecture,
+configuration, and S3A support.  Supporting new committers on older Hadoops
+will be challenging.  It also seems to be the version where the magic output
+committer is recommended for use, so potentially our users will be there or
+will agree to upgrade.
+
+#### ParquetOutputCommitters
+
+Parquet requires its committers to be
+[`ParquetOutputCommitter`](https://github.com/apache/parquet-mr/blob/5608695f5777de1eb0899d9075ec9411cfdf31d3/parquet-hadoop/src/main/java/org/apache/parquet/hadoop/ParquetOutputCommitter.java#L37)s
+(of course it does), see e.g. [Cloudera's
+explanation](https://docs.cloudera.com/HDPDocuments/HDP3/HDP-3.1.5/bk_cloud-data-access/content/enabling-directory-committer-spark.html).
+It's not intended to be derived from, and cannot easily use another
+OutputCommitter.  However there's a
+[`BindingParquetOutputCommitter`](https://github.com/apache/spark/blob/08e6f633b5bc3a7d8d008db2a264b1607d269f25/hadoop-cloud/src/hadoop-3/main/scala/org/apache/spark/internal/io/cloud/BindingParquetOutputCommitter.scala#L37)
+in Spark (not in Hadoop, not in Parquet-MR, in *Spark*) that claims to
+transform the selected output committer into a ParquetOutputCommitter.  This
+is resolvable per documentation but will require some work.
+
+## Future work
+
+### Multiple writers
+
+We can support multiple concurrent writers and allow all writers to succeed
+(keeping the results of the last writer).
+
+We add a "merge-if" operation to lakeFS: atomically merge branch B into
+branch A if branch A is at a given commit.
+
+Now change LakeFSOutputCommitter to loop during commit:
+
+* Attempt to merge the task branch into the source branch _if the source
+  branch has not moved_.
+* On failure:
+  - Merge the source branch into the task branch using the "destination
+	wins" strategy and delete all added files under the prefix (or add and
```

Comment:
Shouldn't we use the "source wins" strategy here? The merge destination is the task branch in this case

Reply:
Not really.  This is a merge from the source to the task.  Whenever there is a conflict, we want the task branch (which will become the "latest writer" after a successful HOC commit) to win.

Added text. :-)